### ISSUE TITLE
MNIST Distributed example is incorrect

### ISSUE THREAD

**Issue by miguelvr** [pytorch/ignite#615]
The MNIST distributed example does not reduce the loss tensors, and therefore is incorrect.

By not reducing the loss, backpropagation is currently being computed on each process separately, leading to different sets of weights.

To fix this, a custom process function with a call to all_reduce needs to be passed to the engine.

Same applies to the metrics computed. They will all be unreliable since none of them is taking into account the distributed scenario.

I would say that this needs to be fixed ASAP, because it is extremely misleading for new users.

**Comment by vfdev-5** [comment]
@miguelvr yes [mnist_dist.py](https://github.com/pytorch/ignite/blob/master/examples/mnist/mnist_dist.py) needs to be reworked or even to be removed. For distributed example I would suggest [cifar10](https://github.com/pytorch/ignite/tree/distrib/examples/contrib/cifar10).

**Comment by miguelvr** [comment]
doesn't seem like the loss is being reduced in that example either

**Comment by vfdev-5** [comment]
Each process has its batch loss and we log only batch loss of the master. IMO synchronization and copying at each iteration can cost (to check).

**Comment by vfdev-5** [comment]
> By not reducing the loss, backpropagation is currently being computed on each process separately, leading to different sets of weights.

I'm not really sure about this. If we take pytorch/examples/imagenet as example of correct distributed training, there is no reduction of loss across batches : https://github.com/pytorch/examples/blob/master/imagenet/main.py#L293

Am I missing something ?

**Comment by miguelvr** [comment]
correct me if i'm wrong but:

if you don't call `loss = torch.distributed.all_reduce(loss)` before the backward call, you are computing different gradients in each process, leading to different model parameter version per process.

I've checked [maskrcnn_benchmark code](https://github.com/facebookresearch/maskrcnn-benchmark/blob/24c8c90efdb7cc51381af5ce0205b23567c3cd21/maskrcnn_benchmark/engine/trainer.py#L76-L85) and they seem to reduce the loss before the `backward` call.

I'm not sure if pytorch handles this "automagically" or not.

**Comment by vfdev-5** [comment]
I would say, that
```
        loss_dict_reduced = reduce_loss_dict(loss_dict)
        losses_reduced = sum(loss for loss in loss_dict_reduced.values())
        meters.update(loss=losses_reduced, **loss_dict_reduced)
```
is only for correct logging and not the training. 

Yes, pytorch handles this "automagically" :) See https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel
```
This container parallelizes the application of the given module by splitting the input across the specified devices by chunking in the batch dimension. The module is replicated on each machine and each device, and each such replica handles a portion of the input. During the backwards pass, gradients from each node are averaged.
```

**Comment by miguelvr** [comment]
ah cool :) 👍 thanks for the clarification
